libraries/protocol/include/graphene/protocol/restriction.hpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2018 Abit More, and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <graphene/protocol/base.hpp>
+#include <graphene/protocol/types.hpp>
+
+namespace graphene { namespace protocol {
+
+/**
+  * Defines the set of valid operation restritions as a discriminated union type.
+  */
+struct restriction {
+   enum function_type {
+      func_eq,
+      func_ne,
+      func_lt,
+      func_le,
+      func_gt,
+      func_ge,
+      func_in,
+      func_not_in,
+      func_has_all,
+      func_has_none,
+      func_attr,
+      func_logical_or,
+      func_variant_assert,
+      FUNCTION_TYPE_COUNT ///< Sentry value which contains the number of different types
+   };
+
+   // A variant assertion argument is a pair of the tag expected to be in the variant, and the restrictions to apply
+   // to the value
+   using variant_assert_argument_type = pair<int64_t, vector<restriction>>;
+
+#define GRAPHENE_OP_RESTRICTION_ARGUMENTS_VARIADIC \
+   /*  0 */ void_t, \
+   /*  1 */ bool, \
+   /*  2 */ int64_t, \
+   /*  3 */ string, \
+   /*  4 */ time_point_sec, \
+   /*  5 */ public_key_type, \
+   /*  6 */ fc::sha256, \
+   /*  7 */ account_id_type, \
+   /*  8 */ asset_id_type, \
+   /*  9 */ force_settlement_id_type, \
+   /* 10 */ committee_member_id_type, \
+   /* 11 */ witness_id_type, \
+   /* 12 */ limit_order_id_type, \
+   /* 13 */ call_order_id_type, \
+   /* 14 */ custom_id_type, \
+   /* 15 */ proposal_id_type, \
+   /* 16 */ withdraw_permission_id_type, \
+   /* 17 */ vesting_balance_id_type, \
+   /* 18 */ worker_id_type, \
+   /* 19 */ balance_id_type, \
```

Comment:
The list was initially generated when HTLC is not implemented, so the data types used in HTLC-related operations E.G. `htlc_id_type` and `htlc_hash` are not in the list. I haven't fully reviewed the code so far, so I'm not sure how users could create custom active authorities (CAA) for the HTLC-related operations. Just think that the CAA feature should be ease to expand. Will update when have more info.

Reply:
The `*_id_type`s are listed here mainly for the convenience to compare 2 containers. For individual id types, it's possible to create restrictions with `func_attr` which will recurse into the `object_id` struct and restrict the `instance` field directly.

Reply:
I was wrong. `object_id::instance` is not reflected, thus we're unable to recurse into it with `func_attr`. It means we need to add the missing types in this type list. Updated the previous comment.

Reply:
Best if we can have code to check whether there is a type missing, or a field that can't define restrictions for. For the fields that aren't supported, we need to document them. In my old implementation (https://github.com/bitshares/bitshares-core/pull/1362/files#diff-e9075b54e174b51fa26c8db6f535b8f2R440) I didn't support all types.

Reply:
> Best if we can have code to check whether there is a type missing, or a field that can't define restrictions for. For the fields that aren't supported, we need to document them. In my old implementation (https://github.com/bitshares/bitshares-core/pull/1362/files#diff-e9075b54e174b51fa26c8db6f535b8f2R440) I didn't support all types.

See 3cd47d1


> The list was initially generated when HTLC is not implemented, so the data types used in HTLC-related operations E.G. `htlc_id_type` and `htlc_hash` are not in the list. I haven't fully reviewed the code so far, so I'm not sure how users could create custom active authorities (CAA) for the HTLC-related operations. Just think that the CAA feature should be ease to expand. Will update when have more info.

See b957a22